### PR TITLE
Expose some more training configuration options.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Expose additional training configuration parameters to Python code.

**Motivation and Context**
Configure number of pre-training graph graph transformer passes and whether GELU approximation is enabled.

### CODE REVIEW

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 1]
```diff
@@ -21,6 +21,9 @@ using namespace onnxruntime::logging;
 using namespace onnxruntime::training;
 
 struct TrainingParameters {
+  std::string model_with_loss_function_path;
```
Comment (edgchen1) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:24]: these are some values i found useful to hardcode in this file for debugging. exposing them to Python would at least allow them to be set without rebuilding ORT, but i'm not sure how/if they should be exposed from ORTTrainer.

File: orttraining/orttraining/python/experimental/orttrainer.py [hunk 2]
```diff
@@ -581,6 +581,9 @@ def _create_ort_training_session(self):
         ort_parameters.weights_to_train = trainable_params
         ort_parameters.optimizer_attributes_map = optimizer_attributes_map
         ort_parameters.optimizer_int_attributes_map = optimizer_int_attributes_map
+        ort_parameters.max_num_pre_training_graph_transformation_steps = \
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/experimental/orttrainer.py:584]: Is it possible to add a unit test for each new flag? How do we check whether it is working?

File: orttraining/orttraining/python/experimental/orttrainer_options.py [hunk 3]
```diff
@@ -217,6 +226,11 @@ class ORTTrainerOptions(object):
             enables gradient norm clipping for 'AdamOptimizer' and 'LambOptimizer'
         utils.invertible_layer_norm_gradient (bool, default is False):
             enables use of invertible layer norm gradients
+        utils.max_num_pre_training_graph_transformation_steps (int, default is 1):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/experimental/orttrainer_options.py:229]: Do you feel `max_num_pre_training_graph_transformation_steps` and `enable_gelu_approximation` really belong to `utils` or another category would be best?

Comment (edgchen1) [reply orttraining/orttraining/python/experimental/orttrainer_options.py:229]: how about a separate one for ORT graph transformer configuration?

File: orttraining/orttraining/python/experimental/orttrainer.py (hunk 2 above)
Comment (edgchen1) [reply orttraining/orttraining/python/experimental/orttrainer.py:584]: added a test, let me know if you had something else in mind

File: orttraining/orttraining/python/experimental/orttrainer_options.py [hunk 4]
```diff
@@ -95,6 +95,22 @@ class ORTTrainerOptions(object):
                         }
                     }
                 },
+                'graph_transformer': {
```
Comment (spandantiwari) [inline orttraining/orttraining/python/experimental/orttrainer_options.py:98]: We should be very deliberate about what goes in to the API. Otherwise we run the risk of bloating the API a lot. Some questions that may help set context to make a decision:
1) Why is this needed? What use cases does this unblock?
2) How is a user expected to take a decision on this? 
3) Is there a way to set this automatically?
4) Is there a similar argument or precedent for this in PyTorch?

**Comment by suffiank** [comment]
> We should be very deliberate about what goes in to the API. Otherwise we run the risk of bloating the API a lot. Some questions that may help set context to make a decision:
> 

Hi @spandantiwari. I agree with your broader concern.

Though I think a couple features would be nice?
1. Ability to toggle specific and/or gross optimization using 'optimization' config field.
2. Ability to export ONNX graph via 'debug' field? (i.e. loss +/- optimization and training +/- optimization).

> Why is this needed? What use cases does this unblock?
At least one optimization, Gelu approximation, lowers the accuracy of the math. So it is off by default on the backend. But in order to match the speed of the BERT C++ run, I had to force it on. It would be nice if we could supply either the optimization level (e.g. 1, 2, 3) and/or specific optimizations (e.g. GeluApproximation, LayerNorm fusion etc.). This is akin to how compiler flags control the assembly outputs.

It is frequently necessary to inspect the model graph in order to understand why the performance does not match expectations. Or to otherwise diagnose the behavior of the running graph. Or to visualize it for other team members.

> How is a user expected to take a decision on this?
> Is there a way to set this automatically?
> Is there a similar argument or precedent for this in PyTorch?
An advanced user would approach optimization fields in the same manner an advanced developer uses compiler optimization flags (c.g. https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html). The compiler also has options to toggle intermediate outputs like preprocessed source file, assembly output etc (= stages of ONNX graph in our case). And even if it is not explicitly documented to end users, it still improves in-house productivity. I'm not aware that Pytorch performs this degree of static analysis.
